File: LintCode/cpp/Pow(x, n).cc
class Solution {
public:
    /**
     * @param x the base number
     * @param n the power number
     * @return the result
     */
    double myPow(double x, int n) {
        if (n < 0) {
            return 1.0 / myPow(x, abs(n));
        }
        if (n == 0) {
            return 1;
        }
        double res = 1;
        if (n & 1) {
            res = x;
        }
        double y = myPow(x, n / 2);
        return res * y * y;
    }
};

File: Leetcode/Algorithm/kotlin/Pow(x, n).kt
class Solution {
    fun myPow(x: Double, n: Int): Double {
        return myPow_(x, n.toLong())
    }
        
    fun myPow_(x: Double, n: Long): Double {
        if (n == 0L) {
            return 1.0
        }
        if (n < 0L) {
            return 1.0 / myPow_(x, -n)
        }
        assert(n > 0);
        
        var res: Double = 1.0
        var xx = x
        var nn = n
        while (nn > 0) {
            // println("${nn}, ${xx}")
            if ((nn and 1L) != 0L) {
                res *= xx
            }
            xx = xx * xx
            nn = nn shr 1
        }
        return res
    }
}
